Sources/Utility/ArgumentParser.swift

```diff
@@ -530,39 +530,44 @@ public final class ArgumentParser {
         ///     simplifications in the parsing code.
         fileprivate func add(_ values: [ArgumentKind], for argument: AnyArgument) throws {
             if argument.isArray {
-                var array = results[argument] as? [ArgumentKind] ?? []
+                var array = results[argument.name] as? [ArgumentKind] ?? []
                 array.append(contentsOf: values)
-                results[argument] = array
+                results[argument.name] = array
             } else {
                 // We expect only one value for non-array arguments.
                 guard let value = values.only else {
                     assertionFailure()
                     return
                 }
-                results[argument] = value
+                results[argument.name] = value
             }
         }
 
         /// Get an option argument's value from the results.
         ///
         /// Since the options are optional, their result may or may not be present.
         public func get<T>(_ argument: OptionArgument<T>) -> T? {
-            return (results[AnyArgument(argument)] as? T) ?? parentResult?.get(argument)
+            return (results[argument.name] as? T) ?? parentResult?.get(argument)
         }
 
         /// Array variant for option argument's get(_:).
         public func get<T>(_ argument: OptionArgument<[T]>) -> [T]? {
-            return (results[AnyArgument(argument)] as? [T]) ?? parentResult?.get(argument)
+            return (results[argument.name] as? [T]) ?? parentResult?.get(argument)
         }
 
         /// Get a positional argument's value.
         public func get<T>(_ argument: PositionalArgument<T>) -> T? {
-            return results[AnyArgument(argument)] as? T
+            return results[argument.name] as? T
         }
 
         /// Array variant for positional argument's get(_:).
         public func get<T>(_ argument: PositionalArgument<[T]>) -> [T]? {
-            return results[AnyArgument(argument)] as? [T]
+            return results[argument.name] as? [T]
+        }
+
+        /// Untyped argument by name
+        public func get(name: String) -> Any? {
```

Comment:
We should add some tests for this new API.

Reply:
I would like this method to be generic, as in:
```swift
public func get<T>(name: String) -> T? {
    return results[name] as? T
}
```

Reply:
Seconded. `Any` in interfaces is a code smell.

Reply:
Sounds good! Also, we should probably throw if there is an entry but the type doesn't match.

Reply:
None of the other getter methods throw an error, I disagree on that point.

Reply:
True but those APIs are completely type-safe. Clients may end up ignoring type changes (for e.g. during a refactoring) silently but if this API throws, then they have to handle the errors.

Reply:
Sorry, not sure I understand. I don't see why the first of these two methods can return optional while the second should throw:

```swift
public func get<T>(_ argument: PositionalArgument<T>) -> T? {
    return results[argument.name] as? T
}

public func get<T>(name: String) -> T? {
    return results[name] as? T
}
```

While the first one receives a PositionalArgument instance, it is not guaranteed it has been added to the same parser. In case of error, I would expect them to fail the same way (return `nil`).

Reply:
I agree that it is possible to create a fake positional argument instance and then try to call that API. However, I don't think the API clients will do that. For the string API, clients may make a mistake and get a type mismatch. Note that I still think it should return optional. Something like this:

```swift
public func get<T>(name: String) throws -> T? {
    if let value = results[name] {
        if let typedValue = value as? T {
              return typedValue
        }
        throw "unexpected type \(type(of: value)) for argument \(name)"
    }
    return nil
}
```

Also, clients can always ignore the errors using `try?`.

We could also make the API contract that only the right type should be accessed. So instead of throwing, we could use assert/precondition but either way I wouldn't want to silently ignore type mismatches. Does that make sense?

Reply:
It does, and I see the value on it. Thanks for explaining!

Reply:
Agree on all points. Should the generic type be an explicit param though, like the Argument initializers (or Codable api's for that matter)?

```swift
public func get<T>(_ name: String, type: T.Type) throws -> T?
```